src/comms.rs

```diff
@@ -620,17 +580,117 @@ pub enum UnitStatus {
     Gone,
 }
 
-impl fmt::Display for UnitStatus {
+impl<'a> From<&'a UnitUpdate> for UnitHealth {
+    fn from(update: &'a UnitUpdate) -> Self {
+        match update {
+            UnitUpdate::Payload(_) => Self::Healthy,
+            UnitUpdate::Stalled => Self::Stalled,
+            UnitUpdate::Gone => Self::Gone,
+        }
+    }
+}
+
+impl fmt::Display for UnitHealth {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         f.write_str(match *self {
-            UnitStatus::Healthy => "healthy",
-            UnitStatus::Stalled => "stalled",
-            UnitStatus::Gone => "gone",
+            UnitHealth::Healthy => "healthy",
+            UnitHealth::Stalled => "stalled",
+            UnitHealth::Gone => "gone",
         })
     }
 }
 
 
+//------------ UnitStatus ----------------------------------------------------
+
+/// A combination of both a unit’s health and the last payload update.
+///
+/// This is a helper type that makes it easier to apply updates.
+#[derive(Clone, Debug, Default)]
+struct UnitStatus {
+    /// The current health of the unit.
+    health: UnitHealth,
+
+    /// The last payload update if there ever was one.
+    payload: Option<payload::Update>,
+}
+
+impl UnitStatus {
+    /// Applies a unit update to the status.
+    ///
+    /// Returns whether the status has changed.
+    fn apply(&mut self, update: &UnitUpdate) -> bool {
```

Comment:
The code in this fn feels like a lot of code for just doing what I think is effectively a compare and set. Can it be reduced to something like (pseudo code) `if self != update { self.set(update); true } else { false }` ?

Reply:
But then you need to impl `PartialEq<UnitUpdate>` and a separate `set` method which both need to match somehow. Not sure that’s really worth it.